ORFS Flow Tutorial (#1559)
The-OpenROAD-Project/OpenROAD: docs/tutorials/FlowTutorial.md

| `VERILOG_FILES`    | The path to the design Verilog files                                                                                                     |
| `SDC_FILE`         | The path to design `.sdc` file                                                                                                           |
| `CORE_UTILIZATION` | The core utilization percentage.                                                                                                         |
| `PL_DENSITY`       | The desired placement density of cells. It reflects how spread the cells would be on the core area. 1 = closely dense. 0 = widely spread |

### Timing Constraints

View timing constraints specified in the `.sdc` file [here](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/blob/master/flow/designs/sky130hd/ibex/constraint.sdc)

```
less ./designs/sky130hd/ibex/constraint.sdc
```

For `ibex` design we simply use the clock definition as follows as a
minimum required timing constraint.

```
create_clock -name core_clock -period 17.4 [get_ports {clk_i}]
```

### Design Input Verilog

The Verilog input files are located in `./designs/src/ibex/`

> The design is defined in `ibex_core.v` availabale [here](https://raw.githubusercontent.com/The-OpenROAD-Project/OpenROAD-flow-scripts/master/flow/designs/src/ibex/ibex_core.v)

Reviewer: why use link to raw.github... instead of www.github like everywhere else?

technology `.lef`, design `.def` files and the parasitics and invokes the
GUI in these steps:

-   Reads and loads `.lef` files (tech and merged)
-   Reads and loads `.def` files (final design def)
-   Loads `.spef` (parasitics)

The figure below shows the post-routed DEF for the `ibex` design.

![ibex_fina_def](./images/ibex_final_def.png)

### Visualizing Design Objects And Connectivity

Note the `Display Control` window on the LHS that shows buttons
for color, visibility and selection options for various design
objects: Layers, Nets, Instances,Blockages, Heatmaps etc.

The Inspector window on the RHS allows user to inspect details of
selected design objects and the timing report.

Try selectively displaying (show/hide) various design objects through
the display control window and observing their impact on the display.

### Tracing The Clock Tree

To view the synthesized clock tree for `ibex` design follow steps:
-   From the top Toolbar Click `View–>Find`
-   In the Dialog Box `Find Object ` search the clock net using keyword as follows:

![cts_find_obect](./images/cts_find_object.png)

Click `Ok` to view the synthesized clock tree of your design.

> To view clock tree structure below, user need to disable the Metal

Reviewer: metal should not be capitalized.

`Layers` section on LHS as shown below.

![ibex_clock_tree](./images/ibex_clock_tree.png)

> From the top Toolbar, Click on  `Windows –> options`

Reviewer: There is no Windows -> options, just a Windows menu

to select/hide different view options of Scripting, Display control etc.

### Using Heat Maps

> From the Menu Bar, Click on `Tools→Heat Maps → Placement Density` to view

Reviewer: Some places you use arrows (like here) and others you use `->`, probably should be consistent about that.
Also there should be a space between Tools and the arrow and the arrow and Heat for consistency.

congestion selectively on vertical and horizontal layers.

To view congestion on all layers between 50-100%

In the `Placement density` setup pop-up window, Select `Minimum → 50.00%`

> `Maximum →100.00%`

Reviewer: space

![congestion_setup](./images/congestion_setup.png)

To view Heat maps for placement density with default settings:

> Expand `Heat Maps →Placement Density` from Display Control window

Reviewer: space and other places

available on LHS of OpenROAD GUI as follows:

![placement_heat_map](./images/placement_heatmap.png)

From `Display Control`, select `Heat Maps →Routing Congestion` as
follows:

![routing_heat_map](./images/routing_heatmap.png)

From `Display Control`, select `Heat Maps →Power Density` as
follows:

![power_heat_map](./images/power_heatmap.png)

### Viewing Timing Report

Click `Timing –> Options` to view and traverse specific timing paths.
From Toolbar click `Timing` icon, View `Timing Report` window added at
right side (RHS) of GUI as shown below.

![Timing report option](./images/timing_report_option.png)

In `Timing Report` Select `Paths → Update`, `Paths` should be integer
numbers. Number of timing paths should be displayed in the current window
as follows:

![Clock Path Update](./images/clock_path_update.png)

Select `Setup` or `Hold` tabs and view required, arrival times, and
slack for each timing path segment.

For each `Setup` or `Hold` path group, path details have a specific `pin

> name, Time, Delay, Slew and Load` value with Clock to register, register

Reviewer: clock

designers. Here user will learn how a DRC violation can be traced with
`gcd` design. Refer to the following openroad test case for more details.

```
cd ./flow/tutorials/scripts/drt/
openroad -gui
```

In the `Tcl Commands` section of GUI run the following command:

```
source drc_issue.tcl
```

Post detail routing in the log you can find number of violations left
in the design:

```
[INFO DRT-0199] Number of violations = 7.
```

Based on `DRC Viewer` steps load `results/5_route_drc.rpt`. GUI as
follows

![gcd DRC issue load](./images/gcd_drc_issue.png)

`X mark` in design highlights DRC violations.

From `DRC Viewer` on RHS `expand > Short`

This shows the number of `violations` in the design. Zoom the design
for clean view on violation:

![View DRC Violation](./images/view_violation.png)

`output53` has overlaps and this causes the `short violation` type.

Open the input DEF file `drc_cts.def` to check the source of the
overlap.

Note the snippet of DEF file where `output51` and `output53` have
the same placed coordinates and hence cause the short violation type.

```
- output51 sky130_fd_sc_hd__clkbuf_1 + PLACED ( 267260 136000 ) N ;

> - output53 sky130_fd_sc_hd__clkbuf_1 + PLACED ( 267260 136000 ) N ;

Reviewer: Wouldn't this have been a placement violation?

and associate the IO cells with those present in the netlist of the
design.

For I/O pad placement using ICeWall refer readme file [here](../../src/pad/README.md)

Refer to the built-in examples [here](../../src/pad/test)

Launch openroad GUI, by typing:

```
cd ../../src/pad/test/
openroad -gui
```

Run [coyote_tc_sky130.tcl](../../src/pad/test/coyote_tc_sky130.tcl) to view ICeWall based IO pad placement.

From the GUI `Tcl commands` prompt run below script:

```
source coyote_tc_sky130.tcl
```

View the resulting IO pad ring in GUI as shown below

![coyote pad ring](./images/coyote_pad_ring.png)

### Power Planning And Analysis

In this section, you will use the design `black_parrot` to create a
power grid and run power analysis.

Pdngen is used for power planning. Refer to the following `README.md` [here](../../src/pdn/README.md)

Refer to the built-in examples [here](../../src/pdn/test)

Launch openroad GUI, by typing:

```
cd ../../src/pdn/test
openroad -gui
```

To run [soc_bsg_black_parrot_nangate45.tcl](../../src/pdn/test/soc_bsg_black_parrot_nangate45.tcl)
to generate power grid for `black_parrot` design by typing:

```
source soc_bsg_black_parrot_nangate45.tcl
```

> Refer to the section of the power domain network strategy file [here](../../src/pdn/test/soc_bsg_black_parrot_nangate45/PDN.cfg)

Reviewer: This is still referencing the PDN config file instead of the correct tcl commands that should be used instead to generate the power grid.

   0.000    0.000   clock network delay (ideal)
   0.000    0.000 ^ r1/CK (DFF_X1)
   0.835    0.835 ^ r1/Q (DFF_X1)
   0.001    0.836 ^ u1/A (BUF_X1)
   0.196    1.032 ^ u1/Z (BUF_X1)
   0.001    1.033 ^ u2/A (BUF_X1)
   0.121    1.154 ^ u2/Z (BUF_X1)
   0.001    1.155 ^ u3/A (BUF_X1)
   0.118    1.273 ^ u3/Z (BUF_X1)
   0.001    1.275 ^ u4/A (BUF_X1)
   0.118    1.393 ^ u4/Z (BUF_X1)
   0.001    1.394 ^ u5/A (BUF_X1)
   0.367    1.761 ^ u5/Z (BUF_X1)
   0.048    1.809 ^ r2/D (DFF_X1)
            1.809   data arrival time

   0.300    0.300   clock clk (rise edge)
   0.000    0.300   clock network delay (ideal)
   0.000    0.300   clock reconvergence pessimism
            0.300 ^ r2/CK (DFF_X1)
  -0.155    0.145   library setup time
            0.145   data required time
-----------------------------------------------------------
            0.145   data required time
           -1.809   data arrival time
-----------------------------------------------------------
           -1.664   slack (VIOLATED)

```

To fix this setup violation use below command:

```
repair_timing -setup
```

The log as follows:

```
[INFO RSZ-0040] Inserted 4 buffers.
[INFO RSZ-0041] Resized 16 instances.

> [WARNING RSZ-0062] Unable to repair all setup violations.

Reviewer: What should the user do if they encounter this? You should probably show how to clean up the design to get proper timing.

Reviewer 2: @vijayank88 Suggest to the user to try lowering the clock frequency and re-run  to resolve the setup violation in this case.

    [OpenROAD-flow-scripts](https://openroad.readthedocs.io/en/latest/user/GettingStarted.html)
    for the complete RTL-to-GDS flow.
-   Users already familiar with the OpenROAD application and flow but
    would like to learn more about specific features and commands.

## User Guide

-   This tutorial requires a specific directory structure built by
    OpenROAD-flow-scripts (ORFS). Do not modify this structure or
    underlying files since this will cause problems in the flow execution.
-   User can run the complete RTL-to-GDS flow and learn specific flow
    sections independently. This allows user to learn the flow and tool
    capabilities at your own pace, time and preference.
-   Results shown such as images or outputs of reports and logs could
    vary based on release updates. However, the main flow and command
    structure should generally apply.

Note: Please submit any problem found under Issues in the GitHub repository
[here](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues)

## Getting Started

This section describes the environment setup to build ORFS and get ready
to execute the RTL-to-GDS flow of the open-source design `ibex` using the

> `sky30hd` technology.

Reviewer 3: ```suggestion
`sky130hd` technology.
```

`ibex` is a 32 bit RISC-V CPU core (`RV32IMC/EMC`) with a two-stage
pipeline.

### Setting Up The Environment

Use the `bash` shell to run commands and scripts.

To install OpenROAD Flow Scripts refer to the [Getting Started with
OpenROAD Flow](../user/GettingStarted.md) documentation.

```
git clone --recursive https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts
cd OpenROAD-flow-scripts
```

If ORFS is already installed but needs updating then use use the
following commands:

```
git checkout master
git pull

> ```

Reviewer 3: It's best to just remove this section and keep info in one place. The second script is missing `git submodule update`.

Author: Thanks @rovinski  as we're already pointing installation links, removing it `git clone`

#### Building OpenROAD Locally

Only the ` --local` build option is used in the tutorial. For further
details refer to [Build Locally](../user/BuildLocally.md) documentation.

Alternatively, user may build using the docker installation by referring
to [Build With Docker](../user/BuildWithDocker.md) documentation.

```
./build_openroad.sh --local
source setup_env.sh
```

> User will see the following message:

Reviewer 3: ```suggestion
You will see the following message:
```

```
OPENROAD: <path>/OpenROAD-flow-scripts/tools/OpenROAD
```

ORFS installation is complete.

## Configuring The Design

This section shows how to set up the necessary platform and design
configuration files to run the complete RTL-to-GDS flow using
`OpenROAD-flow-scripts`.

```
cd flow
```

### Platform Configuration

View the platform configuration file setup for default variables for
`sky130hd`.

```
less ./platforms/sky130hd/config.mk
```

The `config.mk` file has all the required variables for the `sky130`
platform and hence it is not recommended to change any variable

> definition here. To view 
> [sky130hd platform configuration](https://raw.githubusercontent.com/The-OpenROAD-Project/OpenROAD-flow-scripts/master/flow/platforms/sky130hd/config.mk)

Reviewer 3: ```suggestion
definition here. You can view the `sky130hd` platform configuration
[here](https://raw.githubusercontent.com/The-OpenROAD-Project/OpenROAD-flow-scripts/master/flow/platforms/sky130hd/config.mk).
```

Refer to the [Flow variables](../user/FlowVariables.md) document for
details on how to use platform and design specific environment variables
in OpenROAD-flow-scripts to customize and configure your design flow.

### Design Configuration

View the default design configuration of `ibex` design:

```
less ./designs/sky130hd/ibex/config.mk
```

> To view [ibex design config.mk](https://raw.githubusercontent.com/The-OpenROAD-Project/OpenROAD-flow-scripts/master/flow/designs/sky130hd/ibex/config.mk)

Reviewer 3: Don't use raw link.

using Tcl scripts that invoke open sourced tools, from synthesis to the final
`.gds` file creation, without requiring human intervention. In this tutorial,
however, user will learn both the automated and a few interactive ways to
run Tcl commands for important flow stages.

From the OpenROAD-flow-scripts directory user can access individual flow
stages, respective tools and the corresponding `README.md` for tool commands,
configuration examples using the Tcl interface and other such details.

-   Synthesis - [Yosys](https://github.com/The-OpenROAD-Project/yosys/blob/master/README.md)

-   Floorplanning - [Initialize Floorplan](../../src/ifp/README.md)

-   Global Placement - [RePlAce](../../src/gpl/README.md)

-   Clock Tree Synthesis - [TrintonCTS 2.0](../../src/cts/README.md)

-   Detailed Placement - [OpenDP](../../src/dpl/README.md)

-   Global Route - [Fast Route](../../src/grt/README.md)

-   Antenna Rule Checker - [Antenna Rule Checker](../../src/ant/README.md)

-   Timing Optimization using the Resizer - [Gate Resizer](../../src/rsz/README.md)

-   Detail Routing - [TritonRoute](../../src/drt/README.md)

> -   Layout Generation - [KLayout](https://www.klayout.de/) (Requires v0.27.1)

Reviewer 4: remove empty lines

Reviewer: @vijayank88 I meant that you had: `Synthesis-` instead of `Synthesis -` like the test of the table. No need to add lines. It looks like the missing space is resolved.

Reviewer 4: He had asked for single whitespace between the words and the hyphen. Example: `Synthesis-` -> `Synthesis -`

The empty line between a list of items leads to the incorrect rendering of markdown and HTML.

Author: Thanks got it @vvbandeira  @arlpetergadfort

### Design Goals

Run the `ibex` design in ORFS automated flow from RTL-to-GDS using `sky130hd` for given
the design goals below:

-   Area

```
Minimum Required Die size: 0 0 798 800 (in micron)
Core size: 2 2 796 798 (in micron)
```

-   Timing

```
Clock period to meet timing: 17.4 (in ns)
```

> This flow runs for approximately `8 mins` on an `8 core`, `16B RAM` on `CentOS
> 7`. The run time will vary based on your OS and configuration.

Reviewer 3: ```suggestion
This flow takes approximately 8 mins on a machine with 8 cores and 16B RAM.
The runtime will vary based on your configuration.
```

> Change your current directory to the `flow` directory. Type:

Reviewer 3: ```suggestion
Change your current directory to the `flow` directory:
```

```
cd OpenROAD-flow-scripts/flow
```

> To run the complete flow, type below command:

Reviewer 3: ```suggestion
Run the complete flow with:
```

The flow completes with the message below by creating a merged final GDS file.

```
[INFO] Writing out GDS/OAS
'results/sky130hd/ibex/base/6_1_merged.gds'
cp results/sky130hd/ibex/base/6_1_merged.gds
results/sky130hd/ibex/base/6_final.gds
```

## Viewing Results And Logs

The ORFS flow appends a suffix to each flow stage as shown below to
indicate the position in the RTL-GDS flow. This makes it easier to
understand and debug each flow stage in case of failure.

To view `ibex` design logs, in terminal type as:

```
ls logs/sky130hd/ibex/base/
```

The log structure as follows:

```
1_1_yosys.log 2_1_floorplan.log 2_1_floorplan.json 2_2_floorplan_io.log
2_2_floorplan_io.json 2_3_tdms_place.log 2_3_tdms.json
2_4_mplace.log 2_4_mplace.json 2_5_tapcell.log 2_5_tapcell.json 2_6_pdn.log 2_6_pdn.json
3_1_place_gp.log 3_1_place_gp.json 3_2_place_iop.log 3_2_place_iop.json 
3_3_resizer.log 3_3_resizer.json 3_4_opendp.log 3_4_opendp.json
4_1_cts.log 4_1_cts.json 4_2_cts_fillcell.log 4_2_cts_fillcell.json
5_1_fastroute.log 5_1_fastroute.json 5_2_TritonRoute.log
5_2_TritonRoute.json 6_1_merge.log 6_report.log 6_report.json

> ```

Author: @vvbandeira I am using logs table already in this tutorial. how to link same table here too? so we will avoid cloning here

    OpenROAD-flow-scripts (ORFS). Do not modify this structure or
    underlying files since this will cause problems in the flow execution.
-   User can run the full RTL-to-GDS flow and learn specific flow
    sections independently. This allows users to learn the flow and tool
    capabilities at their own pace, time and preference.
-   Results shown, such as images or outputs of reports and logs, could
    vary based on release updates. However, the main flow and command
    structure should generally apply.

Note: Please submit any problem found under Issues in the GitHub repository
[here](https://github.com/The-OpenROAD-Project/OpenROAD-flow-scripts/issues).

## Getting Started

This section describes the environment setup to build ORFS and get ready
to execute the RTL-to-GDS flow of the open-source design `ibex` using the
`sky130hd` technology.

`ibex` is a 32 bit RISC-V CPU core (`RV32IMC/EMC`) with a two-stage
pipeline.

### Setting Up The Environment

Use the `bash` shell to run commands and scripts.

Install OpenROAD Flow Scripts refer to the [Getting Started with
OpenROAD Flow](../user/GettingStarted.md) documentation.

> If ORFS is already installed but needs updating, then use the

Reviewer 3: ```suggestion
If ORFS is already installed but needs updating, run the
```

following commands:

```
cd OpenROAD-flow-scripts
git checkout master
git pull

> git submodule update

Reviewer 3: Why not just `./build_openroad.sh --latest`. This does the same thing as well as also building and installing.

Reviewer 3: Actually please disregard this. This only updates the tools, not the ORFS repository. Please keep as is.

| Variable Name      | Description                                                                                                                              |
|--------------------|------------------------------------------------------------------------------------------------------------------------------------------|
| `PLATFORM`         | Specifies Process design kit.                                                                                                            |
| `DESIGN_NAME`      | The name of the top-level module of the design                                                                                           |
| `VERILOG_FILES`    | The path to the design Verilog files                                                                                                     |
| `SDC_FILE`         | The path to design `.sdc` file                                                                                                           |
| `CORE_UTILIZATION` | The core utilization percentage.                                                                                                         |

> | `PL_DENSITY`       | The desired placement density of cells. It reflects how spread the cells would be on the core area. 1 = closely dense. 0 = widely spread |

Reviewer 3: ```suggestion
| `PLACE_DENSITY`    | The desired placement density of cells. It reflects how spread the cells would be on the core area. 1 = closely dense. 0 = widely spread |
```

## Running The Automated RTL-to-GDS Flow

This section describes the complete execution of the design flow from
RTL-to-GDS. The OpenROAD application executes the entire autonomous flow
using Tcl scripts that invoke open-sourced tools, from synthesis to the final
`.gds` file creation, without requiring human intervention. However, in this
tutorial, the user will learn both the automated and a few interactive ways
to run Tcl commands for important flow stages.

From the OpenROAD-flow-scripts directory, users can access individual flow
stages, respective tools and the corresponding `README.md` for tool commands,
configuration examples using the Tcl interface and other such details.

-   Synthesis - [Yosys](https://github.com/The-OpenROAD-Project/yosys/blob/master/README.md).
-   Floorplanning - [Initialize Floorplan](../../src/ifp/README.md).
-   Global Placement - [RePlAce](../../src/gpl/README.md).
-   Clock Tree Synthesis - [TrintonCTS 2.0](../../src/cts/README.md).
-   Detailed Placement - [OpenDP](../../src/dpl/README.md).
-   Global Route - [Fast Route](../../src/grt/README.md).
-   Antenna Rule Checker - [Antenna Rule Checker](../../src/ant/README.md).

> -   Timing Optimization using the Resizer - [Gate Resizer](../../src/rsz/README.md).

Reviewer 3: ```suggestion
-   Timing Optimization using Resizer - [Gate Resizer](../../src/rsz/README.md).
```

-   Detail Routing - [TritonRoute](../../src/drt/README.md).
-   Layout Generation - [KLayout](https://www.klayout.de/) (Requires v0.27.1).

### Design Goals

Run the `ibex` design in ORFS automated flow from RTL-to-GDS using `sky130hd`

> for a given design goals below:

Reviewer 3: ```suggestion
for the given design goals below:
```

-   Area

```
Minimum Required Die size: 0 0 798 800 (in micron)
Core size: 2 2 796 798 (in micron)
```

-   Timing

```
Clock period to meet timing: 17.4 (in ns)
```

> `ibex` design takes approximately 8 mins on a machine with 8-core and 16GB RAM.

Reviewer 3: ```suggestion
`ibex` takes approximately 8 minutes on a machine with 8-cores and 16GB RAM.
```

The table below briefly describes the reports directory files.

| File Name              | Description                                              |
|------------------------|----------------------------------------------------------|
| `5_route_drc.rpt`      | DRC violations if occurred                               |
| `final_clocks.webp`    | OR extracted image reference after clock tree synthesis. |
| `final_resizer.webp`   | OR extracted image reference after resizer.              |
| `synth_check.txt`      | Synthesis warning/error messages                         |
| `antenna.log`          | Antenna check log report                                 |
| `final_placement.webp` | Extracted image after final placement                    |
| `final.webp`           | Extracted image after routing.                           |
| `synth_stat.txt`       | Post synthesis design statistics log saved here.         |

The flow completes with the message below by creating a merged final GDS file.

```
[INFO] Writing out GDS/OAS
'results/sky130hd/ibex/base/6_1_merged.gds'
cp results/sky130hd/ibex/base/6_1_merged.gds
results/sky130hd/ibex/base/6_final.gds
```

## Viewing Results And Logs

> The ORFS flow appends a suffix to each flow stage, as shown below, to

Reviewer 3: ```suggestion
ORFS prepends a prefix to each flow stage, as shown below, to
```

to [Build With Docker](../user/BuildWithDocker.md) documentation.

```
./build_openroad.sh --local
source setup_env.sh
```

User will see the following message:

```
OPENROAD: <path>/OpenROAD-flow-scripts/tools/OpenROAD
```

ORFS installation is complete.

## Configuring The Design

This section shows how to set up the necessary platform and design
configuration files to run the complete RTL-to-GDS flow using
OpenROAD-flow-scripts.

```
cd flow
```

### Platform Configuration

View the platform configuration file setup for default variables for
`sky130hd`.

```
less ./platforms/sky130hd/config.mk
```

The `config.mk` file has all the required variables for the `sky130`
platform and hence it is not recommended to change any variable
definition here.

![Platform Configuration](./images/platform_configuration.jpg)

Refer to the [Flow variables](../user/FlowVariables.md) document for
details on how to use platform and design specific environment variables
in OpenROAD-flow-scripts to customize and configure your design flow.

### Design Configuration

View the default design configuration of `ibex` design:

```
less ./designs/sky130hd/ibex/config.mk
```

> ![ibex_design_configuration](./images/ibex_design_configuration.png)

Author: yes @arlpetergadfort making link good. No need to do update for minor changes as flow progress.

3_4_opendp.json
4_1_cts.log 4_1_cts.json 4_2_cts_fillcell.log 4_2_cts_fillcell.json
5_1_fastroute.log 5_1_fastroute.json 5_2_TritonRoute.log
5_2_TritonRoute.json 6_1_merge.log 6_report.log 6_report.json
```

### Area

View design area and its core utilization.

```
make gui_final
report_design_area
```

View the resulting area as:

```
Design area 191262 u^2 30% utilization.
```

### Timing

User can view flow results using the command interface from the shell or
the OpenROAD GUI to further visualize, and debug. Learn more about the
[GUI](../../src/gui/README.md)

```
make gui_final
```

Type following commands in `Tcl Commands` section of GUI:

```
report_worst_slack
report_tns
report_wns
```

Note the worst slack, total negative slack and worst negative slack:

```
worst slack -1.35
tns -95.25
wns -1.35
```

To learn more about visualizing and tracing time paths across the design
hierarchy refer to the OpenROAD [GUI](../../src/gui/README.md)

### Power

Now use the report command to view individual power components i.e.
sequential, combinational, macro and power consumed by I/O pads.

```
report_power
```

The power output as follows:

```

> finish report_power

Reviewer: This isn't part of the output

Author: @arlpetergadfort  this is I captured from ibex design 6_report.log.

Reviewer 4: @vijayank88 Peter is right, the `finish report_power` is part of the `final_report.tcl` not the `report_power` command.

Author: report_power comes from final_metrics.tcl
I don't understand what should be updated here @vvbandeira

Author: @vvbandeira  understood, finish report power should not be part of log here

4.  Use heat maps to view congestion and observe the effect of placement
5.  View and trace critical timing paths
6.  Set display control options

If user have completed the RTL-GDS flow, then proceed to view the final
GDS file under results directory `./results/sky130hd/ibex/base/`

For the `ibex` `DESIGN_CONFIG` uncomment the `DESIGN_CONFIG` variable
in the `config.mk`:

```
# DESIGN_CONFIG=./designs/sky130hd/gcd/config.mk
DESIGN_CONFIG=./designs/sky130hd/ibex/config.mk
# DESIGN_CONFIG=./designs/sky130hd/aes/config.mk
```

```
make gui_final
```

### Viewing Layout Results

The `make gui_final` command target successively reads and loads the
technology `.lef`, design `.def` files and the parasitics and invokes the
GUI in these steps:

-   Reads and loads `.lef` files (tech and merged)
-   Reads and loads `.def` files (final design def)
-   Loads `.spef` (parasitics)

The figure below shows the post-routed DEF for the `ibex` design.

![ibex_fina_def](./images/ibex_final_def.png)

### Visualizing Design Objects And Connectivity

Note the `Display Control` window on the LHS (C- Color, V- Visible, S-
Selectable) for various design objects: Layers, Nets, Instances,

> Blockages etc.

Reviewer: C, V, S are no longer in the GUI, need to update the screenshot.

Author: @arlpetergadfort in OpenROAD GUI LHS those icons still exist. But abbreviation not there.
![cvs](https://user-images.githubusercontent.com/85787913/150082016-a345f112-d301-4da5-8821-1a10cf9e731a.png)

Reviewer 2: @vijayank88 , Please update the screenshot and resolve

```
read_lef ./platforms/sky130hd/lef/sky130_fd_sc_hd.tlef
read_lef ./platforms/sky130hd/lef/sky130_fd_sc_hd_merged.lef
read_def ./results/sky130hd/ibex/base/3_place.def
```
![ibex placement density 60](./images/ibex_pl_60.png)

Change `CORE_UTILIZATION` and `PLACE_DENSITY` for the `ibex` design
`config.mk` as follows.

```
export CORE_UTILIZATION = 40
export CORE_ASPECT_RATIO = 1
export CORE_MARGIN = 2
export PLACE_DENSITY = 0.50
```

Re-run the `ibex` design with below command:

```
make DESIGN_CONFIG=./designs/sky130hd/ibex/config.mk
```

View the `ibex` design placement density heat map as shown below:

![ibex placement density 50](./images/ibex_pl_50.png)

So from above GUI understood that change in `CORE_UTILIZATION` from 20
to 40 and placement density default 0.60 to 0.50 changes standard cell
placement became widely spread.

### Chip Level IO Pad Placement

In this section, user will generate an I/O pad ring for the coyote design
using a Tcl script.

ICeWall is a utility to place IO cells around the periphery of a design,
and associate the IO cells with those present in the netlist of the
design.

For I/O pad placement using ICeWall refer readme file below:

```

> less ../../src/pad/README.md

Reviewer: link?

```

Refer to the built-in examples under
`OpenROAD-flow-scripts/tools/OpenROAD/src/ICeWall/test` directory.

Launch openroad GUI, Type

```
cd ../../src/padl/test/
openroad -gui
```

Run `coyote_tc_sky130.tcl` to view ICeWall based IO pad placement.

From the GUI Tcl command prompt run below script,

```
source coyote_tc_sky130.tcl
```

Tcl file reference.

![Coyote IO pad script](./images/coyote_io_pad_script.png)

View the resulting IO pad ring in GUI as shown below

![coyote pad ring](./images/coyote_pad_ring.png)

### Power Planning And Analysis

In this section, user will use the design `black_parrot` to create a
power grid and run power analysis.

Pdngen is used for power planning. Refer to the following `README.md`.

```

> less ../../src/pdn/README.md

Reviewer: link?

```

Refer to the built-in examples under `../../src/pdn/test`
directory.

Launch openroad GUI,

```
cd ../../src/pdn/test
openroad -gui
```

To run `coyote_tc_sky130.tcl` to view ICeWall based IO pad placement
for `black_parrot` design type:

```
source soc_bsg_black_parrot_nangate45.tcl
```

Refer to the section of the power domain network strategy file below:

> ![black_parrot PDN Strategy](./images/blackparrot_pdn.png)

Reviewer: please consider using the tcl interface for PDN for this tutorial as it has improved error checking and is generally more consistent with the rest of the flow.

Reviewer 2: Yes, we will do that
